documentation/specs/proposed/Wasm-tasks.md

```diff
@@ -0,0 +1,344 @@
+    - *-> use JSON strings with callbacks and stdIO for now, with parsing on both sides, WIT is not implemented in wasmtime-dotnet*
+     
+- **TaskExecutionHost?**
+    - TaskExecutionHost is the class that usually runs instantiated tasks and uses reflection to give them property values, 
+    - if we want this layer to handle setting up the environment for the task it has to be abstracted and the interface implemented by custom WasmTaskExecutionHost
+    - Blocked by having to bring the feature to MSBuild repo and refactoring TaskBuilder and including wasmtime-dotnet
+    - *-> keep it separate from MSBuild for now, it's OK that the base WasmTask class will handle setting up the Wasm/WASI environment*
+
+### Related projects
+
+[wasmtime](https://wasmtime.dev/) - Wasm runtime supporting the WASI standard written in Rust by *Bytecode Alliance* - a nonprofit, Microsoft is a member
+
+[wasmtime-dotnet](https://github.com/bytecodealliance/wasmtime-dotnet) - Bindings for wasmtime API in C#
+
+[componentize-dotnet](https://github.com/bytecodealliance/componentize-dotnet) NuGet package to easily make a Wasm/WASI component from a C#.NET project, released short time ago, created by people from Microsoft, right now we can't use it because components are a different system than modules and we can't switch because wasmtime-dotnet does not support it yet.
+
+[dotnet-wasi-sdk](https://github.com/dotnet/dotnet-wasi-sdk) 
+- compile dotnet to Wasm
+- moved to sdk and runtime repos `dotnet workload install wasi-experimental`
+    - Discussions: [1](https://github.com/dotnet/runtime/tree/main/src/mono/wasm) [2](https://github.com/dotnet/runtime/discussions/98538#discussioncomment-8499105) [3](https://github.com/dotnet/runtime/issues/65895#issuecomment-1511265657)
+- developments that would enable using Wasm/WASI tasks written in .NET were added after the workload release but recently removed as .NET will focus only on WASIp2
+
+MSBuild issues for making other environments for running tasks: [711](https://github.com/dotnet/msbuild/issues/711) [4834](https://github.com/dotnet/msbuild/issues/4834) [7257](https://github.com/dotnet/msbuild/issues/7257)
+
+### Random
+
+<!-- https://learn.microsoft.com/en-us/visualstudio/msbuild/configure-tasks?view=vs-2022 -->
+<!-- - configuring tasks to run outside the env of the rest of the project, probably not relevant because wasi is too specific-->
+- wasmtime-dotnet needs to be signed to have a StrongName and put in a private feed if we'd like to integrate it to MSBuild proper eventually https://github.com/bytecodealliance/wasmtime-dotnet/pull/320
+
+- languages other than Rust are moving slowly and it's unclear if they'll implement WASIp1 at all or just like .NET will focus on WASIp2. Closest to support is Go but it's missing function exports. .NET 9 preview 6 possibly implements everything needed for Wasm/WASI tasks but the code was removed in preview 7 and won't come back.
```

Comment:
Ugh 😤. What happened?

Reply:
it was never officially supported and now they switched the Mono WASI effort to target WASIp2  https://github.com/dotnet/runtime/pull/104683

Reply:
Do you have a guess about how much work would be wasted if we used .NET 9-preview5 to build a WASIp1 thing, then converted it to a later WASIp2 system? I would assume _most_ of the .NET work would be transferrable, since it shouldn't directly care about the module/component part, right?

Reply:
To be clear this is about the *guest code* in the Task. ~all of that can't be reused because it'd be generated from the WIT file by https://github.com/bytecodealliance/wit-bindgen which is the point/value of WASIp2 and users would just implement the insides of Execute and GetTaskInfo functions.

Reply:
We transitioned to preview2 and components and we are still in process of adopting it. WASI/dotnet is very experimental.
I'm not sure how easy it is to implement host components (plugins into wasmtime) in dotnet

Reply:
> Do you have a guess about how much work would be wasted if we used .NET 9-preview5 to build a WASIp1 thing

Added value of wit-bindgen is marshaling all the method parameters. 

In p1 you are on your own with memory as array of bytes or you need to use C/rust.
Doing all the marshaling by hands is wasted effort.

Reply:
Yeah but that's a sunk cost with the nice work @JanProvaznik has already done, so that part of the cost should be fairly minimal.